api/agent/drivers/docker/docker.go

```diff
@@ -186,6 +202,80 @@ func containerCleaner(ctx context.Context, driver *DockerDriver) {
 	}
 }
 
+// syncImageCleaner lists the current images on the system and adds them to the
+// image cache. The operation is performed once during startup to ensure a
+// restart of the fn-agent keeps track of previous state.
+func syncImageCleaner(ctx context.Context, driver *DockerDriver) {
+	if driver.imgCache == nil {
+		return
+	}
+
+	const imageListTimeout = time.Duration(60 * time.Second)
+
+	ctx, log := common.LoggerWithFields(ctx, logrus.Fields{"stack": "syncImageCleaner"})
+	limiter := rate.NewLimiter(0.5, 1)
+
+	for limiter.Wait(ctx) == nil {
+		ctx, cancel := context.WithTimeout(ctx, imageListTimeout)
+		images, err := driver.docker.ListImages(docker.ListImagesOptions{Context: ctx})
+		cancel()
+
+		if err == nil {
+			for _, img := range images {
+				log.WithError(err).Errorf("ListImages add %+v", img)
+				driver.imgCache.Update(&CachedImage{
+					ID:       img.ID,
+					ParentID: img.ParentID,
+					RepoTags: img.RepoTags,
+					Size:     uint64(img.Size),
+				})
+			}
+			return
+		}
+
+		log.WithError(err).Error("ListImages error, will retry...")
+	}
+}
+
+// runImageCleaner runs continuously and monitors image cache state. If the
+// cache is over the high water mark limit, then it tries to remove least recently
+// used image.
+func runImageCleaner(ctx context.Context, driver *DockerDriver) {
+	if driver.imgCache == nil {
+		return
+	}
+
+	const removeImgTimeout = time.Duration(60 * time.Second)
+
+	ctx, log := common.LoggerWithFields(ctx, logrus.Fields{"stack": "runImageCleaner"})
+	limiter := rate.NewLimiter(0.5, 1)
+	notifier := driver.imgCache.GetNotifier()
+
+	for limiter.Wait(ctx) == nil {
+		for !driver.imgCache.IsMaxCapacity() {
+			select {
+			case <-ctx.Done(): // driver shutdown
+				return
+			case <-notifier:
+			}
+		}
+
+		img := driver.imgCache.Pop()
+		if img != nil {
+			log.Infof("Removing %+v", img)
+
+			ctx, cancel := context.WithTimeout(ctx, removeImgTimeout)
+			err := driver.docker.RemoveImage(img.ID, docker.RemoveImageOptions{Context: ctx})
+			cancel()
+			if err != nil && err != docker.ErrNoSuchImage {
+				log.WithError(err).Infof("Removing image %+v failed", img)
+				// in-use or can't be removed or docker just timed out, try to add it back to the cache
+				driver.imgCache.Update(img)
```

Comment:
Here the code assumes that the failure is most likely due to the fact that the image is in-use. I went back and forth on this, but Update() seemed OK since it reduces the likelihood of pathological cases such as retrying same image again and again. Head of LRU gives other images a chance to clean up.. We could try to distinguish between various error codes here, but for time being I'm more leaning towards let's observe and see if we run into problems with it. (Originally I had code here that checked timeout condition and placed the image to the back of the LRU, but decided against it later.)